tsinfer/inference.py

```diff
@@ -890,23 +980,77 @@ def __init__(
             [inference_site_position, [sample_data.sequence_length]]
         )
         self.position_map[0] = 0
+        self.recombination = np.zeros(self.num_sites)  # TODO: reduce len by 1
+        self.mismatch = np.zeros(self.num_sites)
+
+        # Set the canonical 'recombination' and 'mismatch' arrays to appropriate values
+        # NB 'recombination' must be of length num_sites-1, mismatch of length num_sites
+
+        # First set defaults if None provided
+        if recombination_rate is None and recombination is None:
+            if mismatch_ratio is None and mismatch is None:
+                # Special case: revert to tsinfer 0.1 behaviour with no mismatch allowed
+                mismatch = np.full(self.num_sites, 1e-20)  # A miniscule value
+                # arbitrary, large relative to mismatch, but < 1
+                if self.num_sites == 0:
+                    recombination = []
+                else:
+                    recombination = np.full(self.num_sites - 1, 1e-2)
+                # Shouldn't need insane precision, although 3 is not enough in trials
+                if precision is None:
+                    precision = 5
+            else:
+                raise ValueError("Cannot use mismatch without setting recombination")
+        if mismatch_ratio is None and mismatch is None:
+            # Default if recombination but no mismatch params have been specified
+            mismatch_ratio = 1.0
+
+        assert not (recombination_rate is None and recombination is None)
+        assert not (mismatch_ratio is None and mismatch is None)
+
+        # Set recombination from recombination_rate
+        if recombination is None:
+            try:
+                recombination = np.diff(
+                    recombination_rate.get_cumulative_mass(inference_site_position)
+                )
+            except AttributeError:
+                recombination = np.diff(inference_site_position * recombination_rate)
+        elif recombination_rate is not None:
+            raise ValueError(
+                "Cannot specify both recombination rate and recombination probabilities"
+            )
+        if len(recombination) != max(0, self.num_sites - 1):
+            raise ValueError("Bad length for recombination array")
+
+        # Set mismatch from mismatch_ratio
+        if mismatch is None:
+            if len(recombination) > 0:
+                mismatch = np.full_like(
+                    self.mismatch, mismatch_ratio * np.median(recombination)
+                )
+            else:
+                # Hack if < 2 sites: https://github.com/tskit-dev/tsinfer/issues/260
+                mismatch = np.full_like(self.mismatch, 0.1)
```

Comment:
Why such a large value? This is basically the same case as pre 0.2 ``mismatch = np.full(self.num_sites, 1e-20)`` above, right?

Reply:
Oh, good point. We could set it to anything, I guess. I assume matching doesn't run anyway if there is only 1 inference site? Or maybe it does run as a special case. This would be run with the recombination probability set to 0, so if it were insanely small we might hit rounding errors if it is run as a special case. But I guess it worked before with recombination=0, mismatch=1e-20?

Reply:
I don't know, but if we're going to pull values out of thin air we might as well minimise the number of different values in case readers think there's some particular significance to them.

Reply:
Good point. I've added (but not pushed yet) a unit test for this, and 1e-20 seems fine, so I'll refactor with that in mind.